File: apps/backend/src/modules/shuttle-optimizer/solvers/or-tools.solver.ts
import { Injectable, Logger } from '@nestjs/common';
import { spawn } from 'node:child_process';
import * as path from 'node:path';
import * as fs from 'node:fs';
import { TSPTWSolver, SolverConfig } from './solver.interface';
import { endDepotMatrixIdx, TSPTWInstance } from '../models/tsptw-instance';
import { TSPTWSolution, emptySolution } from '../models/tsptw-solution';

/**
 * Cấu hình riêng cho OR-Tools solver.
 */
export interface OrToolsConfig extends SolverConfig {
  /** Lệnh Python (mặc định: 'python' trên Windows, 'python3' trên *nix). */
  pythonCommand?: string;
  /** Phạt mỗi vi phạm TW (matches SA). Default 10000. */
  violationPenalty?: number;
}

/** Shape JSON do Python script trả về. */
interface PythonOutput {
  status: 'OK' | 'ERROR';
  route: number[];
  total_distance: number;
  total_duration: number;
  arrival_times: number[];
  violations: number;
  error_message: string | null;
}

/**
 * Google OR-Tools solver — BASELINE CÔNG NGHIỆP.
 *
 * --- Strategy ---
 *
 *   Spawn Python subprocess chạy OR-Tools' Routing Model với
 *   GUIDED_LOCAL_SEARCH metaheuristic. Communication qua stdin/stdout JSON.
 *
 *   Lý do dùng Python: binding Node.js của OR-Tools (`@google/ortools`)
 *   thiếu nhiều feature so với Python wrapper chính thống.
 *
 * --- Pipeline ---
 *
 *   1. NestJS ghi instance JSON ra stdin của Python process
 *   2. `python or-tools-solver.py` đọc, giải, ghi JSON ra stdout
 *   3. NestJS đọc stdout, parse, trả về TSPTWSolution
 *
 * --- Time window xử lý ---
 *
 *   Dùng SOFT upper bound (`SetCumulVarSoftUpperBound`) với penalty lớn.
 *   → Matching với behavior của Greedy / 2-Opt / SA (đếm vi phạm thay vì reject).
 *   → Nếu instance khả thi → OR-Tools sẽ tìm 0-violation.
 *   → Nếu không khả thi → trả nghiệm "ít vi phạm nhất + distance thấp".
 *
 * --- Yêu cầu hệ thống ---
 *
 *   - Python 3.8+
 *   - `pip install ortools` (~150MB)
 *
 *   Nếu thiếu Python hoặc ortools → log warning, trả emptySolution.
 *   Solver khác không bị ảnh hưởng.
 *
 * --- Vai trò trong report ---
 *
 *   Upper bound chất lượng. Optimality gap được tính:
 *
 *     gap = (solver_distance - or_tools_distance) / or_tools_distance × 100%
 *
 *   Kỳ vọng ACO+2opt đạt gap < 5% trên N ≤ 20.
 */
@Injectable()
export class OrToolsSolver extends TSPTWSolver {
  private readonly logger = new Logger(OrToolsSolver.name);

  readonly name = 'or-tools';

  /** Default time limit (ms) — Python subprocess overhead ~1s. */
  private static readonly DEFAULT_TIME_LIMIT_MS = 10_000;

  /** Default violation penalty (matches SA). */
  private static readonly DEFAULT_VIOLATION_PENALTY = 10_000;

  /** Đường dẫn Python script — cùng folder TS file. */
  private readonly scriptPath = path.join(__dirname, 'or-tools-solver.py');

  async solve(
    instance: TSPTWInstance,
    config?: OrToolsConfig,
  ): Promise<TSPTWSolution> {
    const start = Date.now();
    const n = instance.customers.length;

    if (n === 0) {
      return emptySolution(this.name);
    }

    // Kiểm tra script tồn tại
    if (!fs.existsSync(this.scriptPath)) {
      this.logger.error(
        `Không tìm thấy ${this.scriptPath}. ` +
          `Nếu chạy production build, đảm bảo file .py được copy qua nest-cli.json assets.`,
      );
      return emptySolution(this.name);
    }

    // Build input JSON cho Python
    const timeLimitMs =
      config?.timeLimitMs ?? OrToolsSolver.DEFAULT_TIME_LIMIT_MS;
    const violationPenalty =
      config?.violationPenalty ?? OrToolsSolver.DEFAULT_VIOLATION_PENALTY;
    const endIdx = endDepotMatrixIdx(instance);

    const allTimeWindows: Array<[number, number]> = [
      [instance.depot.timeWindow.earliest, instance.depot.timeWindow.latest],
      ...instance.customers.map(
        (c) => [c.timeWindow.earliest, c.timeWindow.latest] as [number, number],
      ),
    ];
    const allServiceTimes: number[] = [
      instance.depot.serviceTime,
      ...instance.customers.map((c) => c.serviceTime),
    ];
    if (instance.endDepot) {
      allTimeWindows.push([
        instance.endDepot.timeWindow.earliest,
        instance.endDepot.timeWindow.latest,
      ]);
      allServiceTimes.push(instance.endDepot.serviceTime);
    }

    const input = {
      distance_matrix: instance.distanceMatrix,
      duration_matrix: instance.durationMatrix,
      time_windows: allTimeWindows,
      service_times: allServiceTimes,
      depot_start: instance.depotStartTime,
      depot_end: instance.depotEndTime,
      end_depot_index: endIdx,
      customer_count: n,
      time_limit_seconds: Math.max(1, Math.floor(timeLimitMs / 1000)),
      violation_penalty: violationPenalty,
    };

    try {
      const output = await this.runPython(
        config?.pythonCommand ?? this.defaultPython(),
        input,
        timeLimitMs + 3000, // overhead 3s cho Python startup
      );
      return this.toSolution(output, Date.now() - start);
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      this.logger.warn(`${this.name} thất bại: ${msg}`);
      return {
        ...emptySolution(this.name),
        runtimeMs: Date.now() - start,
      };
    }
  }

  // ─── Helpers ──────────────────────────────────────────────────────────

  /**
   * Chạy Python subprocess, đẩy `input` qua stdin, đọc JSON từ stdout.
   * Throw nếu script lỗi, exit code khác 0, hoặc timeout.
   */
  private runPython(
    pythonCmd: string,
    input: object,
    killAfterMs: number,
  ): Promise<PythonOutput> {
    return new Promise((resolve, reject) => {
      const proc = spawn(pythonCmd, [this.scriptPath], {
        stdio: ['pipe', 'pipe', 'pipe'],
      });

      let stdout = '';
      let stderr = '';
      let killed = false;

      const killTimer = setTimeout(() => {
        killed = true;
        proc.kill('SIGKILL');
      }, killAfterMs);

      proc.stdout.on('data', (chunk: Buffer) => {
        stdout += chunk.toString('utf8');
      });
      proc.stderr.on('data', (chunk: Buffer) => {
        stderr += chunk.toString('utf8');
      });

      proc.on('error', (err) => {
        clearTimeout(killTimer);
        reject(
          new Error(
            `Spawn '${pythonCmd}' lỗi: ${err.message}. ` +
              `Kiểm tra Python đã cài chưa.`,
          ),
        );
      });

      proc.on('close', (code) => {
        clearTimeout(killTimer);
        if (killed) {
          reject(new Error(`Timeout sau ${killAfterMs}ms (đã kill)`));
          return;
        }
        if (!stdout.trim()) {
          reject(
            new Error(
              `Python không output gì. exit=${code}. stderr: ${stderr.slice(0, 500)}`,
            ),
          );
          return;
        }
        try {
          const parsed = JSON.parse(stdout) as PythonOutput;
          if (parsed.status === 'ERROR') {
            reject(new Error(parsed.error_message ?? 'Python ERROR'));
            return;
          }
          resolve(parsed);
        } catch (e) {
          reject(
            new Error(
              `Parse JSON output thất bại: ${(e as Error).message}. ` +
                `stdout (500c đầu): ${stdout.slice(0, 500)}`,
            ),
          );
        }
      });

      // Gửi input qua stdin
      proc.stdin.write(JSON.stringify(input));
      proc.stdin.end();
    });
  }

  private toSolution(out: PythonOutput, runtimeMs: number): TSPTWSolution {
    return {
      route: out.route,
      totalDistance: out.total_distance,
      totalDuration: out.total_duration,
      isFeasible: out.violations === 0,
      violationCount: out.violations,
      arrivalTimes: out.arrival_times,
      solverName: this.name,
      runtimeMs,
    };
  }

  /** Chọn lệnh Python theo platform — 'python' Win, 'python3' *nix. */
  private defaultPython(): string {
    return process.platform === 'win32' ? 'python' : 'python3';
  }
}

File: apps/backend/src/modules/shuttle-optimizer/solvers/or-tools-solver.py
#!/usr/bin/env python3
"""
OR-Tools TSPTW solver subprocess.

Đọc instance JSON từ stdin, ghi solution JSON ra stdout.

Input format (stdin):
{
  "distance_matrix":  [[float, ...], ...]      # km, [0]=start depot, optional [N+1]=end depot
  "duration_matrix":  [[float, ...], ...]      # phút, cùng kích thước
  "time_windows":     [[earliest, latest], ...] # phút từ 00:00, len = N+1
  "service_times":    [int, ...]               # phút, len = N+1
  "depot_start":      int                      # phút (= time_windows[0][0])
  "depot_end":        int                      # phút (= time_windows[0][1])
  "end_depot_index":  int                      # 0 nếu closed, N+1 nếu open
  "customer_count":   int                      # số customer, không tính depot
  "time_limit_seconds": int                    # giới hạn thời gian solver
  "violation_penalty": int                     # phạt soft TW (default 10000)
}

Output format (stdout):
{
  "status":         "OK" | "ERROR",
  "route":          [int, ...]                 # customer indices (0..N-1)
  "total_distance": float                      # km
  "total_duration": float                      # phút
  "arrival_times":  [float, ...]               # phút, theo route
  "violations":     int                        # số TW bị vi phạm
  "error_message":  str | null
}

Cài đặt:
    pip install ortools

Lưu ý: distance/duration float được scale × 1000 để đưa về integer cho OR-Tools.
"""
import json
import sys
import traceback


def main() -> int:
    try:
        raw = sys.stdin.read()
        data = json.loads(raw)
    except Exception as e:
        return write_error(f"Parse stdin failed: {e}")

    try:
        from ortools.constraint_solver import routing_enums_pb2  # type: ignore
        from ortools.constraint_solver import pywrapcp  # type: ignore
    except ImportError:
        return write_error(
            "ortools chưa cài. Chạy: pip install ortools"
        )

    try:
        return solve(data, pywrapcp, routing_enums_pb2)
    except Exception as e:  # noqa: BLE001
        return write_error(
            f"Solver exception: {e}\n{traceback.format_exc()}"
        )


def write_error(msg: str) -> int:
    """Ghi error JSON ra stdout."""
    out = {
        "status": "ERROR",
        "route": [],
        "total_distance": 0.0,
        "total_duration": 0.0,
        "arrival_times": [],
        "violations": 0,
        "error_message": msg,
    }
    sys.stdout.write(json.dumps(out))
    sys.stdout.flush()
    return 1


def solve(data, pywrapcp, routing_enums_pb2) -> int:
    distance_matrix = data["distance_matrix"]
    duration_matrix = data["duration_matrix"]
    time_windows = data["time_windows"]
    service_times = data["service_times"]
    depot_start = int(data["depot_start"])
    depot_end = int(data["depot_end"])
    end_depot_index = int(data.get("end_depot_index", 0))
    customer_count = int(data.get("customer_count", len(distance_matrix) - 1))
    time_limit_seconds = int(data.get("time_limit_seconds", 5))
    violation_penalty = int(data.get("violation_penalty", 10_000))

    n = len(distance_matrix)  # = N+1 nodes (depot + customers)

    # Scale float → integer cho OR-Tools (× 1000 km = mét, × 1 phút = phút)
    SCALE_DISTANCE = 1000  # km → mét
    distance_int = [
        [int(round(distance_matrix[i][j] * SCALE_DISTANCE)) for j in range(n)]
        for i in range(n)
    ]
    duration_int = [
        [int(round(duration_matrix[i][j])) for j in range(n)] for i in range(n)
    ]

    # 1 vehicle, start depot index 0, end depot có thể khác 0.
    manager = pywrapcp.RoutingIndexManager(n, 1, [0], [end_depot_index])
    routing = pywrapcp.RoutingModel(manager)

    # ─── Distance callback (objective) ──────────────────────────────────
    def distance_cb(from_index, to_index):
        i = manager.IndexToNode(from_index)
        j = manager.IndexToNode(to_index)
        return distance_int[i][j]

    distance_idx = routing.RegisterTransitCallback(distance_cb)
    routing.SetArcCostEvaluatorOfAllVehicles(distance_idx)

    # ─── Time callback (= duration + service) ───────────────────────────
    def time_cb(from_index, to_index):
        i = manager.IndexToNode(from_index)
        j = manager.IndexToNode(to_index)
        return duration_int[i][j] + service_times[i]

    time_idx = routing.RegisterTransitCallback(time_cb)

    # Horizon = big-M. KHÔNG dùng depot_end làm trần cứng: nếu chặn cumul ≤ depot_end
    # thì với instance bị ràng buộc chặt sẽ không tồn tại phép gán nào → OR-Tools
    # trả None ("không tìm được nghiệm"). Đặt trần đủ lớn để SOFT upper bound (đếm
    # vi phạm) điều khiển, đồng bộ hành vi với Greedy/2-opt/SA/ACO.
    max_edge = max((max(row) for row in duration_int), default=0)
    max_service = max(service_times, default=0)
    horizon = depot_start + n * (max_edge + max_service) + depot_end + violation_penalty

    # Time dimension — slack & capacity = horizon (cho phép wait & vi phạm mềm)
    routing.AddDimension(
        time_idx,
        horizon,  # max slack (wait time) — không chặn cứng
        horizon,  # capacity = big-M, KHÔNG phải depot_end (sửa lỗi mô hình)
        False,    # don't force start cumul to 0
        "Time",
    )
    time_dim = routing.GetDimensionOrDie("Time")

    # ─── Time window constraints với SOFT upper bound ──────────────────
    # Lý do soft: matching behavior với Greedy/SA — đếm vi phạm, không từ chối.
    for node in range(1, n):  # bỏ qua depot start/end
        if node == end_depot_index:
            continue
        earliest, latest = time_windows[node]
        index = manager.NodeToIndex(node)
        # Hard lower (đến sớm thì wait, OK)
        time_dim.CumulVar(index).SetMin(int(earliest))
        # Soft upper — vượt qua chỉ bị penalty, không infeasible
        time_dim.SetCumulVarSoftUpperBound(
            index, int(latest), violation_penalty * SCALE_DISTANCE
        )

    # Depot ràng buộc: xuất phát đúng giờ
    depot_index = routing.Start(0)
    time_dim.CumulVar(depot_index).SetRange(depot_start, depot_start)
    # Hạn về depot: trần cứng = horizon (big-M), không chặn ở depot_end.
    end_index = routing.End(0)
    time_dim.CumulVar(end_index).SetMax(horizon)
    # Soft upper tại end → về trễ depot_end chỉ bị phạt, không vô nghiệm.
    time_dim.SetCumulVarSoftUpperBound(
        end_index, depot_end, violation_penalty * SCALE_DISTANCE
    )

    # ─── Search parameters ──────────────────────────────────────────────
    search_params = pywrapcp.DefaultRoutingSearchParameters()
    search_params.first_solution_strategy = (
        routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC
    )
    search_params.local_search_metaheuristic = (
        routing_enums_pb2.LocalSearchMetaheuristic.GUIDED_LOCAL_SEARCH
    )
    search_params.time_limit.seconds = time_limit_seconds

    solution = routing.SolveWithParameters(search_params)

    if solution is None:
        return write_error("OR-Tools không tìm được nghiệm trong time limit")

    # ─── Extract solution ───────────────────────────────────────────────
    route = []
    arrival_times = []
    violations = 0
    total_distance = 0.0
    total_duration_min = 0
    index = routing.Start(0)
    prev_node = None

    while not routing.IsEnd(index):
        node = manager.IndexToNode(index)
        if 1 <= node <= customer_count:
            route.append(node - 1)  # đổi lại về customer index (0..N-1)
            arrival = solution.Value(time_dim.CumulVar(index))
            arrival_times.append(arrival)
            earliest, latest = time_windows[node]
            if arrival > latest:
                violations += 1
        if prev_node is not None:
            total_distance += distance_matrix[prev_node][node]
        prev_node = node
        index = solution.Value(routing.NextVar(index))

    # Leg cuối: prev_node -> depot kết thúc
    end_node = manager.IndexToNode(index)
    if prev_node is not None:
        total_distance += distance_matrix[prev_node][end_node]

    # Total duration = time tại depot end - depot start
    end_time = solution.Value(time_dim.CumulVar(routing.End(0)))
    total_duration_min = end_time - depot_start
    if end_time > depot_end:
        violations += 1  # depot return late

    out = {
        "status": "OK",
        "route": route,
        "total_distance": round(total_distance, 2),
        "total_duration": round(float(total_duration_min), 1),
        "arrival_times": [float(t) for t in arrival_times],
        "violations": violations,
        "error_message": None,
    }
    sys.stdout.write(json.dumps(out))
    sys.stdout.flush()
    return 0


if __name__ == "__main__":
    sys.exit(main())
